### ISSUE TITLE
[apex] Declaration Order

### ISSUE THREAD

**Issue by mitchspano** [pmd/pmd#5431]
**Proposed Rule Name:** [apex] Declaration Order

**Proposed Category:** Code Style

**Description:** This rule enforces a consistent ordering of class members within Apex classes. 
The content in an Apex class needs to be declared in the following order:

1.  static constants
1.  static variables
1.  static methods
1.  instance variables
1.  constructors
1.  instance methods

The elements within each group should be ordered by access modifier (From most open to least open):
1. global
1. public
1. protected
1. private (default)

**Code Sample:**

```java
// Violates the rule: Incorrect order of declarations
public class MyClass {

  public void myMethod() {
    // ...
  }

  private static final Integer MY_CONST = 1;

  public MyClass() {
    // ...
  }

  private String myInstanceVariable;

  public static void myStaticMethod() {
    // ...
  }
}

// Compliant code: Correct order of declarations
public class MyClass {

  public static final Integer MY_CONST = 1;

  public static void myStaticMethod() {
    // ...
  }

  private String myInstanceVariable;

  public MyClass() {
    // ...
  }

  public void myMethod() {
    // ...
  }
}
```

**Possible Properties:**

*  N/A

**Comment by sirephil** [comment]
We do have the basic structuring as you have listed, though for attributes and methods we separate by static/instance - static first with constructors before all other instance methods. Other methods are in whatever order makes sense to the developer (which may collect them by concept, access or even no order whatsoever).

We don't care about method order because most of the time you don't read the code in a linear manner. However, putting statics first, along with attributes, makes those easier to find.

That said, IDEs make code navigation and inspection function regardless of how the code is ordered in the file.

**Comment by pkozuchowski** [comment]
Template Method design pattern may lead to mixed private/protected methods order, in which case order of the usage will be correct.

**Comment by sirephil** [comment]
> 1	Static variables	First the public static variables, then the private.
> 2	Instance variables	First public, then protected, then private.
> 3	Constructors	First public, then protected, then private.
> 4	Methods	These methods should be grouped by functionality rather than by scope or accessibility. For example, a private class method can be in between two public instance methods. The goal is to make reading and understanding the code easier.

Looks good to me. You didn't cover nested classes (but then this was split into a separate rule for some reason).

**Comment by pkozuchowski** [comment]
Both the Google Java Style and the C# rule are talking about order of modifiers on a single variable (ex. `private static final Integer` is good  vs `static private final Integer` is bad). So that circles back to my comment - wiser heads than mine had some reason to omit method order.

- Static Methods Separated - Personally, I think this one is better. Seeing static method mixed with instance method is confusing. However, I think typically static methods should be below constructors?

- Order by Access Modifier - Honestly, not sure. Both sides have some merits. I wouldn't make distinction between protected and private, they should be placed per usage order. There's also another matter of @TestVisible which can also be thought as sort of access modifier.
